File: algorithms/[973]最接近原点的 K 个点/solution.ts
/**
 * 排序
 * time O(nlogn), space O(logn), n 为 points 的长度
 */
function kClosest(points: number[][], k: number): number[][] {
  points.sort(([x1, y1], [x2, y2]) => {
    return x1 * x1 + y1 * y1 - x2 * x2 - y2 * y2;
  });

  return points.slice(0, k);
}

File: algorithms/[973]最接近原点的 K 个点/solution.py
# topics = ["堆"]

import heapq
from typing import List


class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        """
        heap
        time O(nlogk), space O(k), n 为 points 的长度
        """
        q = [(x ** 2 + y ** 2, i) for i, (x, y) in enumerate(points)]
        # 转化为小根堆
        heapq.heapify(q)

        return [points[i] for _, i in heapq.nsmallest(k, q)]

    def kClosest2(self, points: List[List[int]], k: int) -> List[List[int]]:
        """
        heap
        time O(nlogk), space O(k), n 为 points 的长度
        """
        q = [(-(x ** 2) - y ** 2, i) for i, (x, y) in enumerate(points[:k])]
        # 维护长度为 k 的小根堆
        heapq.heapify(q)

        n = len(points)
        for i in range(k, n):
            x, y = points[i]
            dist = -(x ** 2) - y ** 2
            heapq.heappushpop(q, (dist, i))

        ans = [points[i] for _, i in q]
        return ans
